# SQL проект - аналитика онлайн сервиса "Книги"

## Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Моя компания решила быть на волне и купила крупный сервис для чтения книг по подписке. 

## Цель и задачи проекта

__Необходимо проанализировать базы данных для формулировки ценностного предложения для нового продукта.__

__Задачи:__

- Импорт данных, библиотек и вывод первых строк таблиц
- Исследование таблиц — вывести первые строки
- Посчитать, сколько книг вышло после 1 января 2000 года
- Для каждой книги посчитать количество обзоров и среднюю оценку
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры
- Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок
- Общий вывод и рекомендации.

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Импорт данных, библиотек

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''

con=engine.connect()

## Вывод первых строк таблиц и типов данных

In [2]:
# создадим функцию, для отражения первых строк таблиц
def table_info(table):
    
    """ Функция для вывода первых строк таблиц таблиц"""
    
    query = '''SELECT * FROM {} LIMIT 5'''.format(table)
    df = pd.io.sql.read_sql(query, con = engine)
    display (df)
    display (df.info())
    print()

In [3]:
list = ['books','authors', 'ratings', 'reviews', 'publishers']
for i in list:
    print ('\033[1m' + 'Первые строки таблицы:', i)
    print('\033[0m')
    table_info(i)

Первые строки таблицы: books



book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           5 non-null      int64 
 1   author_id         5 non-null      int64 
 2   title             5 non-null      object
 3   num_pages         5 non-null      int64 
 4   publication_date  5 non-null      object
 5   publisher_id      5 non-null      int64 
dtypes: int64(4), object(2)
memory usage: 368.0+ bytes


None


Первые строки таблицы: authors



author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  5 non-null      int64 
 1   author     5 non-null      object
dtypes: int64(1), object(1)
memory usage: 208.0+ bytes


None


Первые строки таблицы: ratings



rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  5 non-null      int64 
 1   book_id    5 non-null      int64 
 2   username   5 non-null      object
 3   rating     5 non-null      int64 
dtypes: int64(3), object(1)
memory usage: 288.0+ bytes


None


Первые строки таблицы: reviews



review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  5 non-null      int64 
 1   book_id    5 non-null      int64 
 2   username   5 non-null      object
 3   text       5 non-null      object
dtypes: int64(2), object(2)
memory usage: 288.0+ bytes


None


Первые строки таблицы: publishers



publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  5 non-null      int64 
 1   publisher     5 non-null      object
dtypes: int64(1), object(1)
memory usage: 208.0+ bytes


None

__Вывод:__ Вывели первые 5 строк всех таблиц базы данных.

## Исследовательский анализ данных

### Посчитать, сколько книг вышло после 1 января 2000 года


In [4]:
# Считаем количество книг, вышедших после 1 января 2000 года
query = '''


SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'


'''

df = pd.io.sql.read_sql(query, con = engine)
print('После 01 января 2020 г. вышло {} книг'.format(df.loc[0, 'count']) )

После 01 января 2020 г. вышло 819 книг


In [5]:
# Считаем общее количество книг
query = '''


SELECT COUNT(book_id)
FROM books

'''

df = pd.io.sql.read_sql(query, con = engine)
print('После 01 января 2020 г. вышло {} книг'.format(df.loc[0, 'count']) )

После 01 января 2020 г. вышло 1000 книг


Всего вышло 1000 книг, из них большая часть - 819 книга вышли после 1 января 2000 года.

### Для каждой книги посчитать количество обзоров и среднюю оценку

In [6]:
query = '''


SELECT b.title,
       b.book_id,
       COUNT (DISTINCT rev.review_id) AS c,
       ROUND (avg(rat.rating), 2) as r
FROM reviews AS rev
RIGHT JOIN books AS b ON rev.book_id=b.book_id
LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
GROUP BY b.book_id
ORDER BY c DESC

'''

df = pd.io.sql.read_sql(query, con = engine)
df.columns = ['Название книги', 'Идентификатор книги', 'Количество обзоров', 'Средняя оценка']
df.head(5)

Название книги  Идентификатор книги  \
0                            Twilight (Twilight  #1)                  948   
1                                Water for Elephants                  963   
2                                   The Glass Castle                  734   
3  Harry Potter and the Prisoner of Azkaban (Harr...                  302   
4  The Curious Incident of the Dog in the Night-Time                  695   

   Количество обзоров  Средняя оценка  
0                   7            3.66  
1                   6            3.98  
2                   6            4.21  
3                   6            4.41  
4                   6            4.08

Вывод: Для каждой книги посчитали количество обзоров и среднюю оценку. Лидер по количеству обзоров - книга Twilight - 7 обзоров со средней оценкой 3.6.

### Определить издательство, которое выпустило наибольшее число книг

Учитываем книги толще 50 страниц, чтобы исключить брошюры

In [7]:
query = '''

SELECT p.publisher,
       p.publisher_id,
       COUNT(b.book_id) AS count_book     
FROM publishers AS p
INNER JOIN books AS b ON p.publisher_id=b.publisher_id
WHERE num_pages > 50
GROUP BY p.publisher_id
ORDER BY count_book DESC
LIMIT 1

'''

df = pd.io.sql.read_sql(query, con = engine)
df.head(5)

publisher  publisher_id  count_book
0  Penguin Books           212          42

Больше всего книг выпустило издательство Penguin Books - 42 книги толще 50 страниц.

### Определить автора с самой высокой средней оценкой книг

Учитываем авторов книг с 50 и более оценками

In [8]:
query = '''

SELECT a.author,
       ROUND(AVG(r.rating),2) avg_rating
FROM authors AS a
INNER JOIN books AS b ON a.author_id=b.author_id
INNER JOIN ratings AS r ON b.book_id=r.book_id
WHERE b.book_id IN
                 (SELECT b.book_id
                  FROM books AS b
                  JOIN ratings AS r ON b.book_id=r.book_id
                  GROUP BY b.book_id
                  HAVING count(r.rating_id)>=50)
GROUP BY a.author
ORDER BY avg_rating DESC
LIMIT 1


'''

df = pd.io.sql.read_sql(query, con = engine)
df.head(5)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

Автор с самой высокой средней оценкой книг J.K. Rowling со средней оценкой - 4.29

### Посчитатть среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [9]:
query = '''

WITH a AS
  (SELECT username,
          count(review_id) AS count_rev
   FROM reviews
   WHERE username in
       (SELECT username
        FROM ratings
        GROUP BY username
        HAVING count(rating_id)>48)
   GROUP BY username)
SELECT round (SUM (count_rev) / COUNT(count_rev), 2) AS avg
FROM a
LIMIT 1


'''

df = pd.io.sql.read_sql(query, con = engine)
print ('Среднее количество отзывов среди самых активных пользователей составляет ', df.loc[0, 'avg'])

Среднее количество отзывов среди самых активных пользователей составляет  24.0


## Общий вывод и рекомендации

__Для формулировки ценностного предложения для онлайн сервиса чтения книг проанализировали базу данных.__

__Получены ответы на поставленные вопросы:__

- Сколько книг вышло после 1 января 2000 года: `819 книг`.
- Для каждой книги определили количество обзоров и среднюю оценку: Лидер - книга `Twilight` - 7 обзоров со средней оценкой 3.6.
- Издательство, которое выпустило наибольшее число книг толще 50 страниц: `Penguin Books` 42 книги
- Автор с самой высокой средней оценкой книг: `J.K. Rowling` со средней оценкой 4.3
- Среднее количество обзоров от пользователей, которые поставили больше 48 оценок: `24`

__Рекомендации:__

Для максимизации дохода и удовлетворенности пользователей необходимо: 

- иметь в доступе все современные бестселлеры
- сделать удобную фильтрацию литературы по темам, жанрам, авторам, рейтингам, рецензиям
- расширить возможные форматы книг: тексты, аудиокники, подкасты к книгам, просто тематические подкасты
- возможно стоит организовать сообщество книголюбов внутри приложения, где можно находить людей по общим интересам.